runtime/databricks/automl_runtime/forecast/pmdarima/training.py

```diff
@@ -65,25 +66,51 @@ def fit(self, df: pd.DataFrame) -> pd.DataFrame:
         # Impute missing time steps
         history_pd = self._fill_missing_time_steps(history_pd, self._frequency_unit)
 
+        history_timedelta = history_pd['ds'].max() - history_pd['ds'].min()
+        timedeltas = history_pd['ds'].diff()
+        min_timedelta = timedeltas.iloc[timedeltas.values.nonzero()[0]].min()
+
         # Tune seasonal periods
         best_result = None
         best_metric = float("inf")
+        all_validation_horizons = dict()
+        all_cutoffs = dict()
         for m in self._seasonal_periods:
             try:
-                cutoffs = generate_cutoffs(history_pd, horizon=self._horizon, unit=self._frequency_unit,
-                                           num_folds=self._num_folds, seasonal_period=m)
+                # this check mirrors the the default behavior by prophet
+                seasonality_timedelta = pd.to_timedelta(m, unit=self._frequency_unit)
+                if history_timedelta < 2 * seasonality_timedelta:
+                    _logger.warning(f"Skipping seasonal_period={m} ({seasonality_timedelta}). Dataframe timestamps must span at least two seasonality periods, but only spans {history_timedelta}")
+                    continue
+                # Prophet also rejects the seasonality periods if the seasonality period timedelta is less than the shortest timedelta in the dataframe.
+                # However, this cannot happen in ARIMA because _fill_missing_time_steps imputes values for each _frequency_unit,
+                # so the minimum valid seasonality period is always 1
+
+                validation_horizon = utils.get_validation_horizon(history_pd, self._horizon, self._frequency_unit)
+                cutoffs = utils.generate_cutoffs(
+                    history_pd,
+                    horizon=validation_horizon,
+                    unit=self._frequency_unit,
+                    num_folds=self._num_folds,
+                )
+                all_validation_horizons[m] = validation_horizon
+                all_cutoffs[m] = cutoffs
+
                 result = self._fit_predict(history_pd, cutoffs, m, self._max_steps)
                 metric = result["metrics"]["smape"]
                 if metric < best_metric:
                     best_result = result
                     best_metric = metric
             except Exception as e:
-                print(f"Encountered an exception with seasonal_period={m}: {repr(e)}")
+                _logger.warning(f"Encountered an exception with seasonal_period={m}: {repr(e)}")
+                traceback.print_exc()
         if not best_result:
             raise Exception("No model is successfully trained.")
 
         results_pd = pd.DataFrame(best_result["metrics"], index=[0])
         results_pd["pickled_model"] = pickle.dumps(best_result["model"])
+        results_pd._validation_horizons = all_validation_horizons
```

Comment:
Yeah I agree perhaps we do not need to record all the horizon and cutoff information. It's not really meant for users to access directly. We only need the information for:
- showing an alert when the validation horizon is different from the user-specified horizon
- testing and debugging
I recorded all the information just in case we need it, eg. to show a more detailed alert or debug trial notebooks.

Reply:
I'm wondering if there is a better way to store this information if it's mostly just for testing and debugging. If we record all the horizons and cutoffs, the `results_pd` can be quite messy, which could make the `arima_results` in the final notebook also messy.
For the alert, I think we only need the final horizon used for the best model. What about only recording the horizon and cutoffs for the best model?

Reply:
Every model should use the same horizons/cutoffs, but different identities may have different horizons/cutoffs. Sure, I can remove this part. Originally, I wanted to pass the horizon/cutoffs from the training results to use for alerting, but it is much easier to just recompute the horizons/cutoffs on the driver for alerting.